src/collapse/perf.rs

```diff
@@ -324,6 +387,91 @@ impl PerfState {
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> result::Result<String, Error> {
+        let addr = u64::from_str_radix(addr, 16).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse {} as hex address, caused by: {}",
+                addr, e
+            ))
+        })?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let bytes = fs::read(module)?;
+            let file = &object::File::parse(&bytes).map_err(|e| {
+                Error::Parse(format!(
+                    "unable to parse file: {}, caused by: {}",
+                    module, e
+                ))
+            })?;
+            entry.insert(addr2line::Context::new(file)?);
+        };
+        let ctx = &self.addr2line_contexts[module];
+
+        let mut frames = ctx.find_frames(addr).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse frames from module: {}, at address: {}. caused by: {}",
+                module, addr, e
+            ))
+        })?;
+
+        let mut full_func = SmallVec::<[String; 1]>::new();
+        let mut total_length = 0;
+        while let Some(frame) = frames.next().unwrap() {
+            let mut function = frame
+                .function
+                .map(|func| {
+                    let name = remove_discriminator(
+                        func.raw_name()
+                            .map(Cow::from)
+                            .unwrap_or_else(|_| Cow::from("??")),
+                    );
+                    strip_rust_hash(addr2line::demangle_auto(name, func.language).as_ref())
+                        .to_string()
+                })
+                .unwrap_or_else(|| "??".into());
+
+            if self.opt.show_context {
+                let loc = frame.location;
+                let (file, line) = match loc {
+                    Some(ref loc) => {
+                        let file = loc
+                            .file
+                            .as_ref()
+                            .map(|f| Path::new(f).file_name())
+                            .map(|f| f.map(OsStr::to_str))
+                            .unwrap_or(None)
+                            .unwrap_or(None);
+                        (file, loc.line)
+                    }
+                    None => (None, None),
+                };
+                function = format!(
+                    "{}:{}:{}",
+                    function,
+                    file.unwrap_or("??"),
+                    line.map(|l| l.to_string()).unwrap_or_else(|| "?".into())
+                );
+            }
+
+            total_length += function.len();
+            full_func.push(function);
+        }
+
+        let mut joined = String::with_capacity(total_length + full_func.len() - 1);
```

Comment:
I think it would actually be better to _not_ join these here, and instead push them all directly onto `self.stack` in the right order. I don't think there's a good reason to join them separately here? Even though I know that's what the Perl version does.

Reply:
Yes, I was mimicking the Perl version, but there's no reason to join here since they'll get joined later if we push them directly to the stack.

Reply:
Just make sure the order stays the same!